src/guide/typescript-support.md

```diff
@@ -0,0 +1,280 @@
+  title: string
+  okMessage: string
+  cancelMessage: string
+}
+const Component = defineComponent({
+  props: {
+    name: String,
+    success: { type: String },
+    callback: {
+      type: Function as PropType<() => void>
+    },
+    message: {
+      type: Object as PropType<ComplexMessage>,
+      required: true,
+      validator(message: ComplexMessage) {
+        return !!message.title
+      }
+    }
+  }
+})
+```
+
+If you find validator not getting type inference or member completion isn’t working, annotating the argument with the expected type may help address these problems.
+
+## Using with Composition API
+
+On `setup()` function, you don't need to pass a typing to `props` parameter as it will infer types from `props` component option.
+
+```ts
+import { defineComponent } from 'vue'
+
+const Component = defineComponent({
+  props: {
+    message: {
+      type: String,
+      required: true
+    }
+  },
+
+  setup(props) {
+    const result = props.message.split('') // correct, 'message' is typed as a string
+    const filtered = props.message.filter(p => p.value) // an error will be thrown: Property 'filter' does not exist on type 'string'
+  }
+})
+```
+
+### Typing `ref`s
+
+Refs infer the type from the initial value:
+
+```ts
+import { defineComponent, ref } from 'vue'
+
+const Component = defineComponent({
+  setup() {
+    const year = ref(2020)
+
+    const result = year.value.split('') // => Property 'filter' does not exist on type 'number'
+  }
+})
+```
+
+Sometimes we may need to specify complex types for a ref's inner value. We can do that simply passing a generic argument when calling ref to override the default inference:
+
+```ts
+const year = ref<string | number>('2020') // year's type: Ref<string | number>
+
+year.value = 2020 // ok!
+```
+
+::: tip Note
+If the type of the generic is unknown, it's recommended to cast `ref` to `Ref<T>`.
+:::
+
+### Typing `reactive`
+
+When typing a `reactive` property, we can use use interfaces:
+
+```ts
+import { defineComponent, reactive } from 'vue'
+
+interface Book {
+  title: string
+  year?: number
+}
+
+export default defineComponent({
+  name: 'HelloWorld',
+  setup() {
+    const book: Book = reactive({ title: 'Vue 3 Guide' })
+    // or
+    const book = reactive({ title: 'Vue 3 Guide' }) as Book
+  }
+})
+```
+
+Reactive should have the same type as the argument passed to `reactive` method. If there's an unwrapped `ref`, within `reactive` object, the types will differ:
+
+```ts
+export default defineComponent({
+  name: 'HelloWorld',
+  setup() {
+    const book: Book = reactive({ title: 'Vue 3 Guide' })
+
+    book.year = ref(2020) // 'Ref<number>' is not assignable to type 'number'
+  }
+})
+```
+
+### Typing `computed`
+
+For computed, we could specify a type with getter only, or with getter and setter:
+
+```ts
+import { defineComponent, ref, computed } from "vue";
+
+export default defineComponent({
+  name: "HelloWorld",
+  setup() {
+    let count = ref(0);
+
```

Comment:
can we have an example with `computed` automatically infer the return type? 

The `read-only` bellow can become override computed type

Reply:
@pikax done!